src/connection_string.ts

```diff
@@ -785,26 +779,842 @@ function parseURI(uri: string): { srv: boolean; url: URL; hosts: string[] } {
     throw new MongoParseError('Invalid connection string, protocol and host(s) required');
   }
 
-  const authString = username ? (password ? `${username}:${password}` : `${username}`) : '';
+  decodeURIComponent(username ?? '');
+  decodeURIComponent(password ?? '');
+
+  const illegalCharacters = new RegExp('[@/:]', 'gi');
+  if (username?.match(illegalCharacters)) {
+    throw new MongoParseError(`Username contains unescaped characters ${username}`);
+  }
+  if (!username || !password) {
+    const uriWithoutProtocol = uri.replace(`${protocol}://`, '');
+    if (uriWithoutProtocol.startsWith('@') || uriWithoutProtocol.startsWith(':')) {
+      throw new MongoParseError('URI contained empty userinfo section');
+    }
+  }
+
+  if (password?.match(illegalCharacters)) {
+    throw new MongoParseError('Password contains unescaped characters');
+  }
+
+  let authString = '';
+  if (typeof username === 'string') authString += username;
+  if (typeof password === 'string') authString += `:${password}`;
+
+  const srv = protocol.includes('srv');
+  const hostList = hosts.split(',');
+  const url = new URL(`${protocol.toLowerCase()}://${authString}@dummyHostname${rest}`);
+
+  if (srv && hostList.length !== 1) {
+    throw new MongoParseError('mongodb+srv URI cannot have multiple service names');
+  }
+  if (srv && hostList[0].includes(':')) {
+    throw new MongoParseError('mongodb+srv URI cannot have port number');
+  }
+
   return {
-    srv: protocol.includes('srv'),
-    url: new URL(`${protocol.toLowerCase()}://${authString}@dummyHostname${rest}`),
+    srv,
+    url,
     hosts: hosts.split(',')
   };
 }
 
+function getBoolean(name: string, value: unknown): boolean {
+  if (typeof value === 'boolean') return value;
+  const valueString = String(value).toLowerCase();
+  const truths = ['true', 't', '1', 'y', 'yes'];
+  const lies = ['false', 'f', '0', 'n', 'no', '-1'];
+  if (truths.includes(valueString)) return true;
+  if (lies.includes(valueString)) return false;
+  throw new TypeError(`For ${name} Expected stringified boolean value, got: ${value}`);
+}
+
+function getInt(name: string, value: unknown): number {
+  if (typeof value === 'number') return Math.trunc(value);
+  const parsedValue = Number.parseInt(String(value), 10);
+  if (!Number.isNaN(parsedValue)) return parsedValue;
+  throw new TypeError(`Expected ${name} to be stringified int value, got: ${value}`);
+}
+
+function getUint(name: string, value: unknown): number {
+  const parsedValue = getInt(name, value);
+  if (parsedValue < 0) {
+    throw new TypeError(`${name} can only be a positive int value, got: ${value}`);
+  }
+  return parsedValue;
+}
+
+function isRecord(value: unknown): value is Record<string, any> {
```

Comment:
I did add the isArray check, how do you wanna reuse it? should this file export the method for the tests to use? that does mean it will have to be exported by our library even if we mark internal.

I don't think its terrible to have a bit of code dupe for this kind of small check.

Reply:
I was just thinking maybe we want to make an `isPlainObject` helper in `src/utils.ts` which could be reused here and in the test file. I agree the cope dupe is minimal here, but we just uncovered a bug in this implementation, can we be sure the two methods will be the same in the future?

Reply:
Yep sounds good, moving to utils makes sense, it is done 👍